acvm-repo/acir/src/circuit/opcodes.rs

```diff
@@ -14,23 +14,85 @@ pub use memory_operation::{BlockId, MemOp};
+    /// ```
+    /// \sum_{i,j} {q_M}_{i,j}w_iw_j + \sum_i q_iw_i +q_c = 0
+    /// ```
+    ///
+    /// An assert-zero opcode can be used to:
+    /// - **express a constraint** on witnesses; for instance to express that a
+    ///   witness `w` is a boolean, you can add the opcode: `w*w-w=0`
+    /// - or, to **compute the value** of an arithmetic operation of some inputs.
+    /// For instance, to multiply two witnesses `x` and `y`, you would use the
+    /// opcode `z-x*y=0`, which would constrain `z` to be `x*y`.
+    ///
+    /// The solver expects that at most one witness is not known when executing the opcode.
     AssertZero(Expression),
-    /// Calls to "gadgets" which rely on backends implementing support for specialized constraints.
+
+    /// Calls to "gadgets" which rely on backends implementing support for
+    /// specialized constraints.
     ///
-    /// Often used for exposing more efficient implementations of SNARK-unfriendly computations.  
+    /// Often used for exposing more efficient implementations of
+    /// SNARK-unfriendly computations.  
+    ///
+    /// All black box functions take as input a tuple `(witness, num_bits)`,
+    /// where `num_bits` is a constant representing the bit size of the input
+    /// witness, and they have one or several witnesses as output.
+    ///
+    /// Some more advanced computations assume that the proving system has an
+    /// 'embedded curve'. It is a curve that cycles with the main curve of the
+    /// proving system, i.e the scalar field of the embedded curve is the base
+    /// field of the main one, and vice-versa. The curves used by the proving
+    /// system are dependent on the proving system (and/or its configuration).
+    ///
+    /// Aztec's Barretenberg uses BN254 as the main curve and Grumpkin as the
+    /// embedded curve.
     BlackBoxFuncCall(BlackBoxFuncCall),
+
+    /// This opcode is a specialization of a Brillig opcode. Instead of having
+    /// some generic assembly code like Brillig, a directive has a hardcoded
+    /// name which tells the solver which computation to do: with Brillig, the
+    /// computation refers to the compiled bytecode of an unconstrained Noir
+    /// function, but with a directive, the computation is hardcoded inside the
+    /// compiler.
+    ///
+    /// Directives will be replaced by Brillig opcodes in the future.
     Directive(Directive),
+
     /// Atomic operation on a block of memory
+    ///
+    /// ACIR is able to address any array of witnesses. Each array is assigned
+    /// an id (BlockId) and needs to be initialized with the MemoryInit opcode.
+    /// Then it is possible to read and write from/to an array by providing the
+    /// index and the value we read/write, as arithmetic expression. Note that
+    /// ACIR arrays all have a known fixed length (given in the MemoryInit
+    /// opcode below)
+    ///
+    /// - predicate: an arithmetic expression that disable the opcode when it is null.
```

Comment:
```suggestion
    /// - predicate: an arithmetic expression that disables execution of the opcode
```
Wdym when it is null?

Reply:
Also "it" is a bit ambiguous here as it can be read as the predicate or the opcode.